### PR TITLE
[Transforms][IPO] Add remarks for ArgumentPromotion and DeadArgumentE…

### PR DESCRIPTION
**Pull request by yonghong-song**
…limination

ArgumentPromotion and DeadArgumentElimination passes may change function signature. This makes bpf tracing difficult since users either not aware of signature change or need to poke into IR or assembly to understand the function signature change.

This patch enabled to emit some remarks so if recompiling with -foptimization-record-file=<file>, users can check remarks to see what kind of signature changes for a particular function. The following are some examples for implemented remarks:
```
  Pass:            deadargelim
  Name:            ReturnValueRemoved
  DebugLoc:        { File: 'bpf-next/net/mptcp/protocol.c', Line: 572, Column: 0 }
  Function:        mptcp_check_data_fin
  Args:
    - String:          'removing return value '
    - String:          '0'

  Pass:            deadargelim
  Name:            ArgumentRemoved
  DebugLoc:        { File: 'bpf-next/kernel/bpf/syscall.c', Line: 1670, Column: 0 }
  Function:        map_delete_elem
  Args:
      - String:          'eliminating argument '
      - ArgName:         uattr.coerce0
      - String:          '('
      - ArgIndex:        '1'
      - String:          ')'

  Pass:            argpromotion
  Name:            ArgumentPromoted
  DebugLoc:        { File: 'bpf-next/net/mptcp/protocol.h', Line: 570, Column: 0 }
  Function:        mptcp_subflow_ctx
  Args:
    - String:          'promoting argument '
    - ArgName:         sk
    - String:          '('
    - ArgIndex:        '0'
    - String:          ')'
    - String:          ' to pass by value'
```
  [1] https://github.com/llvm/llvm-project/issues/104678

### CODE REVIEW

**Comment by yonghong-song** [comment]
cc @4ast @eddyz87 @anakryiko

**Comment by 4ast** [comment]
That should be pretty useful. Is there a tool that can parse it programmatically or, so far, it's for human consumption only?

**Comment by yonghong-song** [comment]
> That should be pretty useful. Is there a tool that can parse it programmatically or, so far, it's for human consumption only?

Alexei, I know there is a meta internal tool to parse remarks to construct an inlining-decision tree. Maybe upstream have some tools to parse those remarks as well. I am not sure.

File: llvm/lib/Transforms/IPO/ArgumentPromotion.cpp [hunk 1]
```diff
@@ -139,6 +141,12 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
     } else if (I->use_empty()) {
       // Dead argument (which are always marked as promotable)
       ++NumArgumentsDead;
+      ORE.emit([&]() {
+        return OptimizationRemark(DEBUG_TYPE, "ArgumentRemoved", F)
+               << ore::NV("Function", F) << " removing argument "
```
Comment (eddyz87) [inline llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:146]: Nitpick: function name is printed twice, first for `OptimizationRemark`, next for `NV`, is it really necessary?

Comment (yonghong-song) [reply llvm/lib/Transforms/IPO/ArgumentPromotion.cpp:146]: Good suggestion. The second function name is not needed. Will fix.

**Review by david-xl — COMMENTED** [review]
Also add a test case?

**Comment by yonghong-song** [comment]
> Also add a test case?

Looks like you are okay with the format? I will add a test case in the next revision.

**Comment by david-xl** [comment]
> > Also add a test case?
> 
> Looks like you are okay with the format? I will add a test case in the next revision.

Adding a case will help get a better feel of the format too.

**Review by david-xl — COMMENTED** [review]
bike shedding on the format. Is it better to show the parameter name first in the message?

foo: eliminating argument 'x' (index 0)
foo: promoting argument 'p'(index 1) to pass by value

**Comment by yonghong-song** [comment]
@david-xl I made some changes to the remark format based on your suggestion. Could you take a look?

**Review by david-xl — APPROVED** [review]
LGTM (you may want to wait a few days in case there are other comments)

**Review by nikic — APPROVED** [review]
Looks reasonable to me.

**Pull request merged**